scio-core/src/main/scala/com/spotify/scio/io/ZstdDictIO.scala

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2024 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.io
+
+import com.github.luben.zstd.ZstdDictTrainer
+import com.spotify.scio.ScioContext
+import com.spotify.scio.coders.CoderMaterializer
+import com.spotify.scio.values.SCollection
+import org.apache.beam.sdk.util.CoderUtils
+import org.slf4j.LoggerFactory
+
+import scala.util.{Random, Try}
+
+case class ZstdDictIO[T](path: String) extends ScioIO[T] {
+  private val logger = LoggerFactory.getLogger(this.getClass)
+
+  override type ReadP = Nothing // WriteOnly
+  override type WriteP = ZstdDictIO.WriteParam
+
+  override val tapT: TapT.Aux[T, Nothing] = EmptyTapOf[T]
+
+  override protected def read(sc: ScioContext, params: Nothing): SCollection[T] =
+    throw new UnsupportedOperationException("ZstdDictIO is write-only")
+
+  override protected def write(
+    data: SCollection[T],
+    params: ZstdDictIO.WriteParam
+  ): Tap[Nothing] = {
+    val ZstdDictIO.WriteParam(
+      zstdDictSizeBytes,
+      numElementsForSizeEstimation,
+      trainingBytesTarget
+    ) = params
```

Comment:
This matching is quite unconventional

Reply:
It's just destructuring ala `case`? I use this frequently but maybe I am an outlier.

Reply:
It's fine. Just noted we've not used this pattern very often